File: forms/post-traumatic-stress-assessment/back-end-with-loco/src/engine/pcl5_grader.rs
//! Pcl5 grader module.

use super::flagged_issues::detect_additional_flags;
use super::pcl5_rules::detect_fired_rules;
use super::types::{AssessmentData, ClusterScores, GradingResult, PclItemScore};
use super::utils::{CUT_OFF, categorise};

/// Sum a slice of PCL item scores, treating `None` as 0.
fn sum(items: &[PclItemScore]) -> i32 {
    items.iter().map(|v| v.unwrap_or(0)).sum()
}

/// Per-cluster sums.
pub fn compute_cluster_scores(d: &AssessmentData) -> ClusterScores {
    ClusterScores {
        b: sum(&[
            d.cluster_b_intrusion.item1_repeated_disturbing_memories,
            d.cluster_b_intrusion.item2_repeated_disturbing_dreams,
            d.cluster_b_intrusion.item3_feeling_reliving,
            d.cluster_b_intrusion.item4_feeling_upset_by_reminders,
            d.cluster_b_intrusion.item5_strong_physical_reactions,
        ]),
        c: sum(&[
            d.cluster_c_avoidance.item6_avoiding_memories_thoughts_feelings,
            d.cluster_c_avoidance.item7_avoiding_external_reminders,
        ]),
        d: sum(&[
            d.cluster_d_negative_alterations.item8_trouble_remembering_important_parts,
            d.cluster_d_negative_alterations.item9_strong_negative_beliefs,
            d.cluster_d_negative_alterations.item10_blaming_self_or_others,
            d.cluster_d_negative_alterations.item11_strong_negative_feelings,
            d.cluster_d_negative_alterations.item12_loss_of_interest,
            d.cluster_d_negative_alterations.item13_feeling_distant_from_others,
            d.cluster_d_negative_alterations.item14_trouble_experiencing_positive_feelings,
        ]),
        e: sum(&[
            d.cluster_e_arousal_reactivity.item15_irritable_or_aggressive,
            d.cluster_e_arousal_reactivity.item16_reckless_or_self_destructive,
            d.cluster_e_arousal_reactivity.item17_super_alert_or_on_guard,
            d.cluster_e_arousal_reactivity.item18_jumpy_or_easily_startled,
            d.cluster_e_arousal_reactivity.item19_difficulty_concentrating,
            d.cluster_e_arousal_reactivity.item20_trouble_sleeping,
        ]),
    }
}

/// Count items at or above the given threshold (treating `None` as 0).
fn count_items_at_or_above(items: &[PclItemScore], threshold: i32) -> i32 {
    items
        .iter()
        .filter(|v| v.unwrap_or(0) >= threshold)
        .count() as i32
}

/// DSM-5 provisional-diagnosis pattern check.
///
/// Requires total >= cut-off AND >= 1 B item, >= 1 C item, >= 2 D items, and
/// >= 2 E items each rated >= 2 (Moderately).
pub fn meets_dsm5_pattern(d: &AssessmentData, total: i32) -> bool {
    if total < CUT_OFF {
        return false;
    }
    let b_vals = [
        d.cluster_b_intrusion.item1_repeated_disturbing_memories,
        d.cluster_b_intrusion.item2_repeated_disturbing_dreams,
        d.cluster_b_intrusion.item3_feeling_reliving,
        d.cluster_b_intrusion.item4_feeling_upset_by_reminders,
        d.cluster_b_intrusion.item5_strong_physical_reactions,
    ];
    let c_vals = [
        d.cluster_c_avoidance.item6_avoiding_memories_thoughts_feelings,
        d.cluster_c_avoidance.item7_avoiding_external_reminders,
    ];
    let d_vals = [
        d.cluster_d_negative_alterations.item8_trouble_remembering_important_parts,
        d.cluster_d_negative_alterations.item9_strong_negative_beliefs,
        d.cluster_d_negative_alterations.item10_blaming_self_or_others,
        d.cluster_d_negative_alterations.item11_strong_negative_feelings,
        d.cluster_d_negative_alterations.item12_loss_of_interest,
        d.cluster_d_negative_alterations.item13_feeling_distant_from_others,
        d.cluster_d_negative_alterations.item14_trouble_experiencing_positive_feelings,
    ];
    let e_vals = [
        d.cluster_e_arousal_reactivity.item15_irritable_or_aggressive,
        d.cluster_e_arousal_reactivity.item16_reckless_or_self_destructive,
        d.cluster_e_arousal_reactivity.item17_super_alert_or_on_guard,
        d.cluster_e_arousal_reactivity.item18_jumpy_or_easily_startled,
        d.cluster_e_arousal_reactivity.item19_difficulty_concentrating,
        d.cluster_e_arousal_reactivity.item20_trouble_sleeping,
    ];
    count_items_at_or_above(&b_vals, 2) >= 1
        && count_items_at_or_above(&c_vals, 2) >= 1
        && count_items_at_or_above(&d_vals, 2) >= 2
        && count_items_at_or_above(&e_vals, 2) >= 2
}

/// Count the number of PCL items that have been answered (non-None).
pub fn answered_item_count(d: &AssessmentData) -> i32 {
    let all = [
        d.cluster_b_intrusion.item1_repeated_disturbing_memories,
        d.cluster_b_intrusion.item2_repeated_disturbing_dreams,
        d.cluster_b_intrusion.item3_feeling_reliving,
        d.cluster_b_intrusion.item4_feeling_upset_by_reminders,
        d.cluster_b_intrusion.item5_strong_physical_reactions,
        d.cluster_c_avoidance.item6_avoiding_memories_thoughts_feelings,
        d.cluster_c_avoidance.item7_avoiding_external_reminders,
        d.cluster_d_negative_alterations.item8_trouble_remembering_important_parts,
        d.cluster_d_negative_alterations.item9_strong_negative_beliefs,
        d.cluster_d_negative_alterations.item10_blaming_self_or_others,
        d.cluster_d_negative_alterations.item11_strong_negative_feelings,
        d.cluster_d_negative_alterations.item12_loss_of_interest,
        d.cluster_d_negative_alterations.item13_feeling_distant_from_others,
        d.cluster_d_negative_alterations.item14_trouble_experiencing_positive_feelings,
        d.cluster_e_arousal_reactivity.item15_irritable_or_aggressive,
        d.cluster_e_arousal_reactivity.item16_reckless_or_self_destructive,
        d.cluster_e_arousal_reactivity.item17_super_alert_or_on_guard,
        d.cluster_e_arousal_reactivity.item18_jumpy_or_easily_startled,
        d.cluster_e_arousal_reactivity.item19_difficulty_concentrating,
        d.cluster_e_arousal_reactivity.item20_trouble_sleeping,
    ];
    all.iter().filter(|v| v.is_some()).count() as i32
}

/// Pure function: grades a PCL-5 assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let cluster_scores = compute_cluster_scores(data);
    let total_score = cluster_scores.b + cluster_scores.c + cluster_scores.d + cluster_scores.e;
    let category = categorise(total_score);
    let probable_dsm5_diagnosis = meets_dsm5_pattern(data, total_score);
    let answered_count = answered_item_count(data);

    let fired_rules = detect_fired_rules(data, total_score, &category, probable_dsm5_diagnosis);
    let additional_flags =
        detect_additional_flags(data, total_score, probable_dsm5_diagnosis, answered_count);

    GradingResult {
        total_score,
        category,
        probable_dsm5_diagnosis,
        cluster_scores,
        fired_rules,
        additional_flags,
        answered_count,
        timestamp,
    }
}

File: forms/post-traumatic-stress-assessment/front-end-with-html/js/pcl5-grader.js
// PCL-5 grader. Pure functions: take an `AssessmentData` object, return the
// total score (0-80), the cluster sub-totals, the `SeverityCategory`, and a
// boolean for the DSM-5 provisional-diagnosis pattern.
//
// Severity bands (PCL-5 total score, 0-80):
//   0-20    -> Minimal
//   21-32   -> Mild
//   33-37   -> Moderate (probable PTSD threshold)
//   38-80   -> Severe
//
// DSM-5 provisional pattern: ≥ 1 B item, ≥ 1 C item, ≥ 2 D items, ≥ 2 E items
// each rated ≥ 2 (Moderately), AND total ≥ 33.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').ClusterScores} ClusterScores
 * @typedef {import('./types.js').SeverityCategory} SeverityCategory
 * @typedef {import('./types.js').PclItemScore} PclItemScore
 */

const CUT_OFF = 33;

/** @param {PclItemScore[]} items */
function sum(items) {
  return items.reduce((acc, v) => acc + (v == null ? 0 : v), 0);
}

/**
 * Per-cluster sums.
 * @param {AssessmentData} d
 * @returns {ClusterScores}
 */
function computeClusterScores(d) {
  return {
    b: sum([
      d.clusterBIntrusion.item1RepeatedDisturbingMemories,
      d.clusterBIntrusion.item2RepeatedDisturbingDreams,
      d.clusterBIntrusion.item3FeelingReliving,
      d.clusterBIntrusion.item4FeelingUpsetByReminders,
      d.clusterBIntrusion.item5StrongPhysicalReactions
    ]),
    c: sum([
      d.clusterCAvoidance.item6AvoidingMemoriesThoughtsFeelings,
      d.clusterCAvoidance.item7AvoidingExternalReminders
    ]),
    d: sum([
      d.clusterDNegativeAlterations.item8TroubleRememberingImportantParts,
      d.clusterDNegativeAlterations.item9StrongNegativeBeliefs,
      d.clusterDNegativeAlterations.item10BlamingSelfOrOthers,
      d.clusterDNegativeAlterations.item11StrongNegativeFeelings,
      d.clusterDNegativeAlterations.item12LossOfInterest,
      d.clusterDNegativeAlterations.item13FeelingDistantFromOthers,
      d.clusterDNegativeAlterations.item14TroubleExperiencingPositiveFeelings
    ]),
    e: sum([
      d.clusterEArousalReactivity.item15IrritableOrAggressive,
      d.clusterEArousalReactivity.item16RecklessOrSelfDestructive,
      d.clusterEArousalReactivity.item17SuperAlertOrOnGuard,
      d.clusterEArousalReactivity.item18JumpyOrEasilyStartled,
      d.clusterEArousalReactivity.item19DifficultyConcentrating,
      d.clusterEArousalReactivity.item20TroubleSleeping
    ])
  };
}

/**
 * @param {number} total
 * @returns {SeverityCategory}
 */
function categorise(total) {
  if (total <= 20) return 'Minimal';
  if (total <= 32) return 'Mild';
  if (total <= 37) return 'Moderate';
  return 'Severe';
}

/** @param {PclItemScore[]} items @param {number} threshold */
function countItemsAtOrAbove(items, threshold) {
  return items.filter((v) => (v == null ? 0 : v) >= threshold).length;
}

/**
 * DSM-5 provisional diagnosis pattern.
 * @param {AssessmentData} d
 * @param {number} total
 */
function meetsDsm5Pattern(d, total) {
  if (total < CUT_OFF) return false;
  const bVals = Object.values(d.clusterBIntrusion);
  const cVals = Object.values(d.clusterCAvoidance);
  const dVals = Object.values(d.clusterDNegativeAlterations);
  const eVals = Object.values(d.clusterEArousalReactivity);
  const b = countItemsAtOrAbove(bVals, 2) >= 1;
  const c = countItemsAtOrAbove(cVals, 2) >= 1;
  const dd = countItemsAtOrAbove(dVals, 2) >= 2;
  const e = countItemsAtOrAbove(eVals, 2) >= 2;
  return b && c && dd && e;
}

/** Count answered PCL items (any of the 20 with a non-null value). */
function answeredItemCount(d) {
  let count = 0;
  const groups = [
    d.clusterBIntrusion,
    d.clusterCAvoidance,
    d.clusterDNegativeAlterations,
    d.clusterEArousalReactivity
  ];
  for (const g of groups) {
    for (const v of Object.values(g)) {
      if (v !== null && v !== undefined) count++;
    }
  }
  return count;
}

/**
 * @param {AssessmentData} d
 * @returns {{ totalScore: number, category: SeverityCategory,
 *             probableDsm5Diagnosis: boolean, clusterScores: ClusterScores,
 *             answeredCount: number }}
 */
function calculatePcl5(d) {
  const clusterScores = computeClusterScores(d);
  const totalScore = clusterScores.b + clusterScores.c + clusterScores.d + clusterScores.e;
  const category = categorise(totalScore);
  const probableDsm5Diagnosis = meetsDsm5Pattern(d, totalScore);
  const answeredCount = answeredItemCount(d);
  return { totalScore, category, probableDsm5Diagnosis, clusterScores, answeredCount };
}

export { CUT_OFF, computeClusterScores, categorise, meetsDsm5Pattern, answeredItemCount, calculatePcl5 };
